src/core/namespace.c

```diff
@@ -2245,32 +2245,42 @@ int setup_namespace(const NamespaceParameters *p, char **error_path) {
         if (r < 0)
                 return r;
 
-        if (p->tmp_dir) {
-                bool ro = streq(p->tmp_dir, RUN_SYSTEMD_EMPTY);
-
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
-                        return log_oom_debug();
-
-                *me = (MountEntry) {
-                        .path_const = "/tmp",
-                        .mode = ro ? MOUNT_PRIVATE_TMP_READ_ONLY : MOUNT_PRIVATE_TMP,
-                        .source_const = p->tmp_dir,
-                };
-        }
+        if (p->private_tmp_as_tmpfs) {
+                r = append_tmpfs_mounts(
+                                &ml,
+                                (TemporaryFileSystem []){{(char *) "/tmp", (char *) "mode=01777" NESTED_TMPFS_LIMITS},
+                                                         {(char *) "/var/tmp", (char *) "mode=01777" NESTED_TMPFS_LIMITS}},
+                                /* n= */ 2);
+                if (r < 0)
+                        return r;
```

Comment:
to make me happy you should be able to just mount /tmp like you do here, but then make /var/tmp/ a bind mount to /tmp/.var/ or so.

That way you maintain this as one superblock, but in separate dirs, and iplementation is detail.